lib/multi_server_socket.dart

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:async/async.dart';
+
+import 'src/utils.dart';
+
+/// The error code for an error caused by a port already being in use.
+final _addressInUseErrno = _computeAddressInUseErrno();
+int _computeAddressInUseErrno() {
+  if (Platform.isWindows) return 10048;
+  if (Platform.isMacOS) return 48;
+  assert(Platform.isLinux);
+  return 98;
+}
+
+/// An implementation of `dart:io`'s [ServerSocket] that wraps multiple servers
+/// and forwards methods to all of them.
+///
+/// This is useful for listening on multiple network interfaces while still
+/// having a unified way of controlling the servers. In particular, it supports
+/// serving on both the IPv4 and IPv6 loopback addresses using
+/// [MultiServerSocket.loopback].
+class MultiServerSocket extends StreamView<Socket> implements ServerSocket {
+  /// The wrapped servers.
+  final Set<ServerSocket> _servers;
+
+  /// Returns the port that one of the wrapped servers is listening on.
+  ///
+  /// If the wrapped servers are listening on different ports, it's not defined
+  /// which port is returned.
+  int get port => _servers.first.port;
+
+  /// Returns the address that one of the wrapped servers is listening on.
+  ///
+  /// If the wrapped servers are listening on different addresses, it's not
+  /// defined which address is returned.
+  InternetAddress get address => _servers.first.address;
+
+  /// Creates a [MultiServerSocket] wrapping [servers].
+  ///
+  /// None of the servers should be listened to when this is called.
+  MultiServerSocket(Iterable<ServerSocket> servers)
+      : _servers = servers.toSet(),
+        super(StreamGroup.merge(servers));
+
+  /// Creates a [ServerSocket] listening on all available loopback addresses for
+  /// this computer.
+  ///
+  /// See [ServerSocket.bind].
+  static Future<ServerSocket> loopback(int port,
+      {int backlog, bool v6Only: false, bool shared: false}) {
+    backlog ??= 0;
+
+    return _loopback(port, 5, backlog, v6Only, shared);
```

Comment:
should the number of retries be configurable?

Reply:
Maybe? But it feels kinda YAGNI to me. I'd rather add it later if a user has a concrete need.

Reply:
Sgtm